File: packages/python/src/cloud_portable_s3vectors/datagen.py
"""Python port of the vector generated-data reference.

Materializes a vector's named datasets and computes the derived digest strings
that ``${data.<name>.<field>}`` placeholders resolve to. Stdlib-only: CRC-32C
and CRC-64/NVME use pure-Python tables (slow on large datasets; the digest
algorithms are size-independent).
"""

from __future__ import annotations

import base64
import hashlib
import struct
import zlib
from collections.abc import Mapping
from typing import Any

__all__ = ["DERIVED_FIELDS", "generate", "derived"]

#: The fields available as ``${data.<name>.<field>}`` placeholders.
DERIVED_FIELDS: tuple[str, ...] = (
    "size", "md5", "etag", "sha256", "sha256B64", "sha1B64",
    "crc32B64", "crc32cB64", "crc64nvmeB64",
)


def _prng(seed: str, size: int) -> bytes:
    # block(i) = SHA256(UTF8(seed) || BE64(i)); stream = block(0) || block(1) || ...
    seeded = hashlib.sha256(seed.encode("utf-8"))
    out = bytearray()
    i = 0
    while len(out) < size:
        h = seeded.copy()
        h.update(struct.pack(">Q", i))
        out += h.digest()
        i += 1
    return bytes(out[:size])


def _pattern(pattern_bytes: bytes, size: int) -> bytes:
    if not pattern_bytes:
        raise ValueError("empty pattern")
    reps = size // len(pattern_bytes) + 1
    return (pattern_bytes * reps)[:size]


def generate(specs: Mapping[str, Mapping[str, Any]], name: str) -> bytes:
    """Materialize one named dataset from a vector's ``data`` map."""
    spec = specs.get(name)
    if spec is None:
        raise KeyError(f"unknown dataset: {name}")
    if "$prng" in spec:
        d = spec["$prng"]
        return _prng(d["seed"], d["size"])
    if "$pattern" in spec:
        d = spec["$pattern"]
        if "pattern" in d:
            return _pattern(d["pattern"].encode("utf-8"), d["size"])
        return _pattern(base64.b64decode(d["patternBase64"]), d["size"])
    if "$slice" in spec:
        d = spec["$slice"]
        parent = specs.get(d["of"])
        if parent is None:
            raise KeyError(f"slice {name!r} references unknown dataset {d['of']!r}")
        if "$slice" in parent:
            raise ValueError(f"slice {name!r} references slice {d['of']!r} (chained slices are not allowed)")
        base = generate(specs, d["of"])
        offset, length = d["offset"], d["length"]
        if offset + length > len(base):
            raise ValueError(
                f"slice {name!r} [{offset}, {offset + length}) exceeds {d['of']!r} size {len(base)}"
            )
        return base[offset:offset + length]
    raise ValueError(f"unknown data kind: {sorted(spec)}")


def _make_crc_table(poly: int, width: int) -> list[int]:
    mask = (1 << width) - 1
    table = []
    for n in range(256):
        c = n
        for _ in range(8):
            c = (c >> 1) ^ poly if c & 1 else c >> 1
        table.append(c & mask)
    return table


_CRC32C_TABLE = _make_crc_table(0x82F63B78, 32)
# CRC-64/NVME: reflected poly, init/xorout all-ones.
_CRC64NVME_TABLE = _make_crc_table(0x9A6C9329AC4BC9B5, 64)


def _crc32c(data: bytes) -> int:
    c = 0xFFFFFFFF
    for b in data:
        c = _CRC32C_TABLE[(c ^ b) & 0xFF] ^ (c >> 8)
    return c ^ 0xFFFFFFFF


def _crc64nvme(data: bytes) -> int:
    c = 0xFFFFFFFFFFFFFFFF
    for b in data:
        c = _CRC64NVME_TABLE[(c ^ b) & 0xFF] ^ (c >> 8)
    return c ^ 0xFFFFFFFFFFFFFFFF


def derived(specs: Mapping[str, Mapping[str, Any]], name: str, field: str) -> str:
    """Compute the string a ``${data.<name>.<field>}`` placeholder resolves to."""
    data = generate(specs, name)
    if field == "size":
        return str(len(data))
    if field == "md5":
        return hashlib.md5(data).hexdigest()
    if field == "etag":
        return f'"{hashlib.md5(data).hexdigest()}"'
    if field == "sha256":
        return hashlib.sha256(data).hexdigest()
    if field == "sha256B64":
        return base64.b64encode(hashlib.sha256(data).digest()).decode("ascii")
    if field == "sha1B64":
        return base64.b64encode(hashlib.sha1(data).digest()).decode("ascii")
    if field == "crc32B64":
        return base64.b64encode(struct.pack(">I", zlib.crc32(data) & 0xFFFFFFFF)).decode("ascii")
    if field == "crc32cB64":
        return base64.b64encode(struct.pack(">I", _crc32c(data))).decode("ascii")
    if field == "crc64nvmeB64":
        return base64.b64encode(struct.pack(">Q", _crc64nvme(data))).decode("ascii")
    raise KeyError(f"unknown derived data field: {field}")

File: packages/go/datagen/datagen.go
// Package datagen is the Go port of the vector generated-data reference:
// it materializes a vector's named datasets and computes the derived digest
// strings that ${data.<name>.<field>} placeholders resolve to.
package datagen

import (
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"hash/crc32"
	"hash/crc64"
	"strconv"

	s3vectors "github.com/alanshaw/s3vectors/packages/go"
)

// DerivedFields lists the fields available as ${data.<name>.<field>} placeholders.
var DerivedFields = []string{
	"size", "md5", "etag", "sha256", "sha256B64", "sha1B64", "crc32B64", "crc32cB64", "crc64nvmeB64",
}

var (
	crc32cTable = crc32.MakeTable(crc32.Castagnoli)
	// CRC-64/NVME reflected polynomial; Go's crc64 applies the all-ones
	// init/xorout of the NVME parameterization internally.
	crc64Table = crc64.MakeTable(0x9A6C9329AC4BC9B5)
)

// block(i) = SHA256(UTF8(seed) || BE64(i)); stream = block(0) || block(1) || ...
func prng(seed string, size int64) []byte {
	out := make([]byte, size)
	seedBytes := []byte(seed)
	var counter [8]byte
	for off, i := int64(0), uint64(0); off < size; off, i = off+32, i+1 {
		binary.BigEndian.PutUint64(counter[:], i)
		h := sha256.New()
		h.Write(seedBytes)
		h.Write(counter[:])
		copy(out[off:], h.Sum(nil))
	}
	return out
}

func pattern(patternBytes []byte, size int64) ([]byte, error) {
	if len(patternBytes) == 0 {
		return nil, fmt.Errorf("empty pattern")
	}
	out := make([]byte, size)
	for off := int64(0); off < size; off += int64(len(patternBytes)) {
		copy(out[off:], patternBytes)
	}
	return out, nil
}

// Generate materializes one named dataset from a vector's data map.
func Generate(specs map[string]s3vectors.DataSpec, name string) ([]byte, error) {
	spec, ok := specs[name]
	if !ok {
		return nil, fmt.Errorf("unknown dataset: %s", name)
	}
	switch {
	case spec.Prng != nil:
		return prng(spec.Prng.Seed, spec.Prng.Size), nil
	case spec.Pattern != nil:
		d := spec.Pattern
		if d.Pattern != nil {
			return pattern([]byte(*d.Pattern), d.Size)
		}
		raw, err := base64.StdEncoding.DecodeString(*d.PatternBase64)
		if err != nil {
			return nil, fmt.Errorf("dataset %s: bad patternBase64: %w", name, err)
		}
		return pattern(raw, d.Size)
	case spec.Slice != nil:
		d := spec.Slice
		parent, ok := specs[d.Of]
		if !ok {
			return nil, fmt.Errorf("slice %q references unknown dataset %q", name, d.Of)
		}
		if parent.Slice != nil {
			return nil, fmt.Errorf("slice %q references slice %q (chained slices are not allowed)", name, d.Of)
		}
		base, err := Generate(specs, d.Of)
		if err != nil {
			return nil, err
		}
		if d.Offset+d.Length > int64(len(base)) {
			return nil, fmt.Errorf("slice %q [%d, %d) exceeds %q size %d",
				name, d.Offset, d.Offset+d.Length, d.Of, len(base))
		}
		return base[d.Offset : d.Offset+d.Length], nil
	default:
		return nil, fmt.Errorf("dataset %s: no $prng/$pattern/$slice key", name)
	}
}

// Derived computes the string a ${data.<name>.<field>} placeholder resolves to.
func Derived(specs map[string]s3vectors.DataSpec, name, field string) (string, error) {
	bytes, err := Generate(specs, name)
	if err != nil {
		return "", err
	}
	switch field {
	case "size":
		return strconv.Itoa(len(bytes)), nil
	case "md5":
		sum := md5.Sum(bytes)
		return hex.EncodeToString(sum[:]), nil
	case "etag":
		sum := md5.Sum(bytes)
		return `"` + hex.EncodeToString(sum[:]) + `"`, nil
	case "sha256":
		sum := sha256.Sum256(bytes)
		return hex.EncodeToString(sum[:]), nil
	case "sha256B64":
		sum := sha256.Sum256(bytes)
		return base64.StdEncoding.EncodeToString(sum[:]), nil
	case "sha1B64":
		sum := sha1.Sum(bytes)
		return base64.StdEncoding.EncodeToString(sum[:]), nil
	case "crc32B64":
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], crc32.ChecksumIEEE(bytes))
		return base64.StdEncoding.EncodeToString(b[:]), nil
	case "crc32cB64":
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], crc32.Checksum(bytes, crc32cTable))
		return base64.StdEncoding.EncodeToString(b[:]), nil
	case "crc64nvmeB64":
		var b [8]byte
		binary.BigEndian.PutUint64(b[:], crc64.Checksum(bytes, crc64Table))
		return base64.StdEncoding.EncodeToString(b[:]), nil
	default:
		return "", fmt.Errorf("unknown derived data field: %s", field)
	}
}
